### PR TITLE
Jquery datable compatible taxonomy endpoint implementation

### LINKED ISSUE #27: New endpoint for Taxonomy API
I attempted to take the outputs of the taxonomy API and produce a basic `DataTable` in javascript for searching / sorting at different levels by the taxa represented in a sample. The outputs are not well suited for this.

What I think would be great is to add in a new endpoint for Taxonomy that allows for the retrieval of JSON structured data that can be fed into a `.DataTable()` constructor. I think it needs to look something like `{"columns": ['foo', 'bar', ...], "data": [[per, row, values, ...]]}`

### PR DESCRIPTION
**Pull request by gwarmstrong**
Fixes #27 . Implements #33 .

This adds the guts for the new taxonomy endpoint. Has potential for forward compatibility with differing taxonomy formats.

Also introduces a `DataTable` class and `create_data_entry` utility that should hopefully aid future development of endpoints that rely on formatting a js DataTable.

### CODE REVIEW

File: microsetta_public_api/api/tests/test_integration.py [hunk 1]
```diff
@@ -171,6 +174,13 @@ def setUp(self):
                                          ['feature-3', 'a; f; g; h', 0.678]],
                                         columns=['Feature ID', 'Taxon',
                                                  'Confidence'])
+        self.taxonomy_greengenes_df = pd.DataFrame(
+            [['feature-1', 'k__a; p__b; o__c', 0.123],
```
Comment (gwarmstrong) [inline microsetta_public_api/api/tests/test_integration.py:178]:
```suggestion
            [['feature-1', 'k__a;p__b; o__c', 0.123],
```

File: microsetta_public_api/models/_taxonomy.py [hunk 2]
```diff
@@ -206,3 +208,64 @@ def get_group(self, ids: Iterable[str], name: str = None) -> GroupTaxonomy:
...
+
+
+class GreengenesFormatter(Formatter):
+    _map = OrderedDict(k__='Kingdom', p__='Phylum', c__='Class',
+                       o__='Order', f__='Family', g__='Genus', s__='Species')
+    labels = list(_map.values())
+
+    @classmethod
+    def dict_format(cls, taxonomy_string: str):
+        ranks = taxonomy_string.split('; ')
```
Comment (gwarmstrong) [inline microsetta_public_api/models/_taxonomy.py:263]:
```suggestion
        ranks = [rank_.strip() for rank_ in taxonomy_string.split(';')]
```

File: microsetta_public_api/models/tests/test_taxonomy.py [hunk 3]
```diff
@@ -36,8 +37,13 @@ def setUp(self):
                                          columns=['Feature ID', 'Taxon',
                                                   'Confidence'])
         self.taxonomy2_df.set_index('Feature ID', inplace=True)
+        self.taxonomy_greengenes_df = pd.DataFrame(
+            [['feature-1', 'k__a; p__b; o__c', 0.123],
+             ['feature-2', 'k__a; p__b; o__c; f__d; g__e', 0.34],
```
Comment (gwarmstrong) [inline microsetta_public_api/models/tests/test_taxonomy.py:42]:
```suggestion
             ['feature-2', 'k__a; p__b; o__c;f__d;g__e', 0.34],
```

File: microsetta_public_api/api/taxonomy.py [hunk 4]
```diff
@@ -50,8 +56,23 @@ def resources():
...
+
+def _present_microbes_taxonomy_table(sample_ids, resource):
+    taxonomy_repo = TaxonomyRepo()
+    error_response = _check_resource_and_missing_ids(taxonomy_repo,
+                                                     sample_ids, resource)
+    if error_response:
+        return error_response
+
+    taxonomy_ = taxonomy_repo.model(resource)
+    taxonomy_table = taxonomy_.presence_data_table(sample_ids)
```
Comment (gwarmstrong) [inline microsetta_public_api/api/taxonomy.py:76]:
```suggestion
    taxonomy_table = taxonomy_repo.presence_data_table(sample_ids)
```

File: microsetta_public_api/models/_taxonomy.py [hunk 5]
```diff
@@ -206,3 +208,64 @@ def get_group(self, ids: Iterable[str], name: str = None) -> GroupTaxonomy:
...
+                                   ).astype('object')
+        sample_data[pd.isna(sample_data)] = None
+        return DataTable.from_dataframe(sample_data)
+
+
+class Formatter:
+
+    labels: List
+
+    @abstractmethod
```
Comment (gwarmstrong) [inline microsetta_public_api/models/_taxonomy.py:251]:
```suggestion
    @abstractclassmethod
```

File: microsetta_public_api/models/_taxonomy.py [hunk 6]
```diff
@@ -1,12 +1,14 @@
-from collections import namedtuple
-from typing import Iterable, Dict
+from collections import namedtuple, OrderedDict
+from typing import Iterable, Dict, Optional, List
+from abc import abstractmethod
```
Comment (gwarmstrong) [inline microsetta_public_api/models/_taxonomy.py:3]:
```suggestion
from abc import abstractclassmethod
```

File: microsetta_public_api/api/taxonomy.py (hunk 4 above)
Comment (gwarmstrong) [reply microsetta_public_api/api/taxonomy.py:76]: Actually, the current usage is fine

File: microsetta_public_api/models/_taxonomy.py [hunk 7]
```diff
@@ -206,3 +219,60 @@ def get_group(self, ids: Iterable[str], name: str = None) -> GroupTaxonomy:
...
+        table = self._table.filter(set(ids), inplace=False).remove_empty()
+        features = table.ids(axis='observation')
+
+        entries = list()
+        for vec, sample_id, _ in table.iter(dense=False):
+            for feature_idx, val in zip(vec.indices, vec.data):
+                entries.append({
+                    **{'sampleId': sample_id,
+                       'relativeAbundance': val},
+                    **self._formatted_taxa_names[features[feature_idx]],
```
Comment (wasade) [inline microsetta_public_api/models/_taxonomy.py:233]: I'm not sure I understand what this does. Possible to rewrite it? I think the following is consistent

```python
entry = {'sampleId': sample_id, 'relativeAbundance', val}
entry.update(self._formatted_taxa_names[features[feature_idx]])
entries.append(entry)
```

File: microsetta_public_api/utils/_utils.py [hunk 8]
```diff
@@ -19,3 +21,64 @@ def check_missing_ids(missing_ids, alpha_metric, type_):
...
+    ----------
+
+    data : list of DataEntry
+        The entries that comprise the DataTable
+    columns : list of str
+        The names of the columns in the DataTable
+
+    """
+    def to_dict(self):
+        data = self.data.copy()
```
Comment (wasade) [inline microsetta_public_api/utils/_utils.py:46]: it looks like this function is first doing a full copy, and then walking back over that same data? I think the amount of effort needed here could be effectively halved with:

```python
dict_ = self._asdict()
dict_['data'] = [entry.to_dict() for entry in enumerate(self.data)]
return dict_
```

File: microsetta_public_api/utils/_utils.py [hunk 9]
```diff
@@ -19,3 +21,64 @@ def check_missing_ids(missing_ids, alpha_metric, type_):
...
+        for i, entry in enumerate(data):
+            # this hinges on the entries of data being DataEntry, since we
+            # know they have a .to_dict() method
+            data[i] = entry.to_dict()
+        dict_ = self._asdict()
+        dict_['data'] = data
+        return dict_
```
Comment (gwarmstrong) [inline microsetta_public_api/utils/_utils.py:53]:
```suggestion
        dict_ = self._asdict()
        dict_['data'] = [entry.to_dict() for entry in
                         enumerate(self.data)]
        return dict_
```

File: microsetta_public_api/api/tests/implementation_tests/test_taxonomy.py [hunk 10]
```diff
@@ -429,3 +431,123 @@ def test_taxonomy_from_list_summarize_single_sample_simple(self):
...
+                    'rank_2': ['b', 'f'],
+                })
+            )
+            response, code = single_sample_taxa_present(
+                'sample-1', "some-table")
+
+        self.assertEqual(code, 200)
+        exp_keys = ['data', 'columns']
+        obs = json.loads(response)
+        print(obs)
```
Comment (gwarmstrong) [inline microsetta_public_api/api/tests/implementation_tests/test_taxonomy.py:495]:
```suggestion
```

File: microsetta_public_api/api/tests/implementation_tests/test_taxonomy.py [hunk 11]
```diff
@@ -429,3 +431,123 @@ def test_taxonomy_from_list_summarize_single_sample_simple(self):
...
+                    'rank_2': ['b', 'f', 'b'],
+                })
+            )
+            response, code = group_taxa_present(
+                {'sample_ids': ['sample-1', 'sample-2']}, "some-table")
+
+        self.assertEqual(code, 200)
+        exp_keys = ['data', 'columns']
+        obs = json.loads(response)
+        print(obs)
```
Comment (gwarmstrong) [inline microsetta_public_api/api/tests/implementation_tests/test_taxonomy.py:530]:
```suggestion
```

File: microsetta_public_api/models/_taxonomy.py [hunk 12]
```diff
@@ -206,3 +219,60 @@ def get_group(self, ids: Iterable[str], name: str = None) -> GroupTaxonomy:
...
+                })
```
Comment (gwarmstrong) [inline microsetta_public_api/models/_taxonomy.py:234]: What if I meet you halfway?
```suggestion
                entry = {
                    'sampleId': sample_id,
                    'relativeAbundance': val,
                    **self._formatted_taxa_names[features[feature_idx]],
                }
                entries.append(entry)
```

**Pull request merged**